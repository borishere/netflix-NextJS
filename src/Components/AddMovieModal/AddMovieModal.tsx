import React, { FC } from 'react';
import { Formik } from 'formik';
import { ImovieBase, validationSchema } from '../../Models/models';
import { useAddMovieMutation } from '../../services/movies';
import { Modal, ModalProps } from '../Modal/Modal';
import { ModalForm } from '../ModalForm/ModalForm';
import styles from './AddMovieModal.module.scss';

export const defaultMovie: ImovieBase = {
  title: '',
  overview: '',
  genres: [],
  vote_average: 0,
  runtime: 0,
  release_date: '',
  poster_path: ''
};

export const AddMovieModal: FC<ModalProps> = ({ ...props }) => {
  const [addMovie, {isLoading}] = useAddMovieMutation();

  const onSubmit = (values: ImovieBase): void => {
    addMovie(values).then(() => {
      props.show(false);
    });
  };

  return (
    <Modal modalClass='add-movie-modal' {...props}>
      <>
        {isLoading && <span>Adding...</span>}
        <h2 className={styles['modal-title']}>ADD MOVIE</h2>
        <Formik<ImovieBase>
          initialValues={defaultMovie}
          validationSchema={validationSchema}
          onSubmit={onSubmit}
        >
          <ModalForm />
        </Formik>
      </>
    </Modal>
  );
};