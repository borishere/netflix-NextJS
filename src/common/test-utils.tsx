import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };