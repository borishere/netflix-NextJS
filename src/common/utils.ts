export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';