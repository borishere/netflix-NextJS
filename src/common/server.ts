import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
