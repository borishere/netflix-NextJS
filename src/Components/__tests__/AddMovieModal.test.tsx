import React from 'react';
import { screen, customRender } from '../../common/test-utils';
import { AddMovieModal } from '../AddMovieModal/AddMovieModal';
import '@testing-library/jest-dom';

describe('AddMovieModal', () => {
  test('is visible', () => {
    const { asFragment } = customRender(
      <AddMovieModal
        isShown={true}
        show={jest.fn()}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('ADD MOVIE')).toBeInTheDocument();
  });

  test('is hidden', () => {
    customRender(
      <AddMovieModal
        isShown={false}
        show={jest.fn()}
      />
    );

    expect(screen.queryByText('ADD MOVIE')).not.toBeInTheDocument();
  });
});