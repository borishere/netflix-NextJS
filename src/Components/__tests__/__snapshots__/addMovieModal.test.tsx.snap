// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieModal is visible 1`] = `
<DocumentFragment>
  <div
    class="modal add-movie-modal"
  >
    <div
      class="modal-content"
    >
      <img
        aria-label="close"
        class="modal-close-btn"
        src="test-file-stub"
      />
      <div
        class="modal-body"
      >
        <h2
          class="modal-title"
        >
          ADD MOVIE
        </h2>
        <form
          action="#"
          class="modal-form"
        >
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                TITLE
              </label>
              <input
                aria-label="title-input"
                class="modal-form__item-input"
                name="title"
                type="text"
                value=""
              />
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RELEASE DATE
              </label>
              <input
                aria-label="date-input"
                class="modal-form__item-input"
                name="release_date"
                placeholder="Select Date"
                type="date"
                value=""
              />
            </div>
          </div>
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                MOVIE URL
              </label>
              <input
                aria-label="url-input"
                class="modal-form__item-input"
                name="poster_path"
                placeholder="https://"
                type="text"
                value=""
              />
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RATING
              </label>
              <input
                aria-label="rating-input"
                class="modal-form__item-input"
                name="vote_average"
                placeholder="7.8"
                type="number"
                value="0"
              />
            </div>
          </div>
          <div
            class="modal-form__item-wrap"
          >
            <div
              class="modal-form__item long"
            >
              <label>
                GENRE
              </label>
              <select
                aria-label="genres-select"
                class="modal-form__item-input"
                multiple=""
                name="genres"
                placeholder="Select Genre"
              >
                <option
                  value="Comedy"
                >
                  Comedy
                </option>
                <option
                  value="Drama"
                >
                  Drama
                </option>
                <option
                  value="Romance"
                >
                  Romance
                </option>
              </select>
            </div>
            <div
              class="modal-form__item"
            >
              <label>
                RUNTIME
              </label>
              <input
                aria-label="runtime-input"
                class="modal-form__item-input"
                name="runtime"
                placeholder="minutes"
                type="number"
                value="0"
              />
            </div>
          </div>
          <div
            class="modal-form__item"
          >
            <label>
              OVERVIEW
            </label>
            <textarea
              aria-label="overview-input"
              class="modal-form__item-input"
              name="overview"
              placeholder="movie description"
              rows="6"
            />
          </div>
          <div
            class="modal-buttons"
          >
            <button
              class="reset-btn"
              type="reset"
            >
              RESET
            </button>
            <button
              class="submit-btn"
              type="submit"
            >
              SUBMIT
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
